### PR TITLE
Initial xformers implementation.

### PR DESCRIPTION
**Pull request by ninele7**
Added initial code for xformers.

Checks xformers installation and applies custom attention if installed.

Also updated pytorch and cudatoolkit (couldn't get it to work without it).

Also provided `install-xformers.sh` to build xformers locally (tested on linux).

(Reminder: xformers must be built for matching GPU-arch, python and cuda versions).

### CODE REVIEW

**Comment by ninele7** [comment]
In this repository: https://github.com/ninele7/xfromers_builds, I will provide action to build wheels for windows and linux for pascal, turing and ampere GPUs.

**Comment by db0** [comment]
Cheers. Please also explain how we can use it as you finalize :)

**Comment by ninele7** [comment]
On that run: https://github.com/ninele7/xfromers_builds/actions/runs/3351726935 I successfully built xformers for windows and linux. 

Can someone test my PR with that on Turing/Ampere GPU? 

To use it (before I add script to download and install it automatically) you need to download artifact for your OS (ubuntu works on any linux distro) and run `./runtime pip install %unpacked wheel name%`.

Than by modifying `test.py` (to load your model and do txt2img/img2img), you can check, that you got performance improvements (on my gpu gains are larger for high-res pictures (768x768 and larger)).

**Comment by db0** [comment]
>Can someone test my PR with that on Turing/Ampere GPU? 

Where are those found?

**Comment by ninele7** [comment]
In actions artifacts there is multiple files:

image

You need to use my PR, download file matching your OS and install it into your conda runtime.

Than It should just work.

**Comment by db0** [comment]
No I meant, how do you know if you have one of those GPUs :D

**Comment by ninele7** [comment]
T4, A100, Geforce RTX 2000/3000. Are ampere and turing.

**Comment by ninele7** [comment]
Marked it as ready. 

Now you can use install-xformers.sh/.cmd to download my build of xformers for Turing/Ampere GPU. After that it should just work.

**Comment by ninele7** [comment]
It's probably useful to add info about xformers install to https://github.com/db0/AI-Horde/blob/main/README_StableHorde.md.

I can make PR after this gets tested by someone other than me.

**Comment by db0** [comment]
OK so explain to someone who doesn't know as much about this tech, like me: This is only beneficial for ampere/turing or all GPUs? Can you explain what this PR achieves exactly?

**Comment by ninele7** [comment]
TLDR: +25-50% performance, slight memory usage reduction without visible image changes. To get benefit need to have Ampere/Turing GPU and run `install-xformers.sh`/`install-xformers.cmd`.

It uses alternative mechanism for CrossAttention provided by xformers (https://github.com/facebookresearch/xformers/) library. On Ampere/Turing gpus it provides 25-50% performance uplift (more noticeable on larger images). I didn't have a chance to test it on Pascal GPU but as far as I heard it doesn't provide much benefit.

To use xformers library it needs to be built for specific python version, os version and cuda arch (can have multiple cuda arches in same binary). `install-xformers` scripts able to install version of xformers for python 3.8 (which conda installs into this repo) and matching os, so users doesn't need to built xformers by themselves.

Also I updated libraries to newer versions because it was easier to make it work with newer versions. (shouldn't impact any users).

**Comment by ninele7** [comment]
More info about xformers available in description for this MR: https://github.com/huggingface/diffusers/pull/532.

**Comment by db0** [comment]
OK so this won't have any benefit on consumer GPUs, ye? I'll ask around if anyone has one of those GPUs and can test

**Comment by ninele7** [comment]
It has benefit on RTX 3000 and RTX 2000 consumer GPUs. I personally tested it on RTX 3060 ti.

**Comment by db0** [comment]
OK, lemme test it on my end

**Comment by db0** [comment]
Question, will this repo work if someone does not build xformers first? I.e. will it just not try to load xformers if they're not built?

**Comment by ninele7** [comment]
In `attention.py` you can see loading logic. It will try to load, but if unsuccessful it will fallback to default implementation.

**Comment by db0** [comment]
Using install-xformers.cmd fails and also breaks my micromamba env 

```ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchvision 0.12.0 requires torch==1.11.0, but you have torch 1.13.0 which is incompatible.
Successfully installed mypy-extensions-0.4.3 pyre-extensions-0.0.23 torch-1.13.0 typing-inspect-0.8.0 xformers-0.0.14.dev0```

I am now trying via `update-runtime.cmd`

**Comment by ninele7** [comment]
Ye, you need to update runtime before running.

**Comment by db0** [comment]
Doesn't this mean xformers will never be loaded?

```python
try:
    importlib_metadata.version("xformers")
except importlib_metadata.PackageNotFoundError:
    xformers_available = False
xformers_available = False
```

**Comment by ninele7** [comment]
ops, I left this debug code, when measured performance difference, you should remove last `xformers_available = False`.

**Comment by db0** [comment]
Thought so. Continuing

**Comment by db0** [comment]
Looks like it's working. Is there any reason why we cannot consoldate everything it into  update-runtime.cmd?

I added the `xformers-0.0.14.dev0-cp38-cp38-win_amd64.whl` into requirements and run `update-runtime.cmd` and it worked fine. Doing it via pip breaks the conda env

**Comment by db0** [comment]
It's better because otherwise it causes a lot of headaches. If you know how to detect GPU cuda capabilitie just tell me and I'll do it. I'm googling it as well in case I can figure it out myself

**Comment by ninele7** [comment]
`(7, 0) <= torch.cuda.get_device_capability(shared.device) <= (9, 0)` Something like that should work.

**Comment by db0** [comment]
checking

**Comment by db0** [comment]
Why is it supposed to be lower than  9?

**Comment by ninele7** [comment]
I haven't tested it on 4090, so safer option would be 8.6 as upper limit, but I saw some people reports, that it should work.

**Comment by db0** [comment]
Hmm, problem exists when trying to have it on requirements.txt , as the name differs per OS...

**Comment by ninele7** [comment]
Would something like that work in requirements.txt? 

```
PKGNAME_1; platform_system == "Windows"
PKGNAME_2; platform_system == "Linux"
```

**Comment by db0** [comment]
Checking

**Comment by db0** [comment]
Looks like it's working. See anything problematic before I merge?

**Comment by ninele7** [comment]
It seems strange to me to use `numba` to get device capability. We have no direct dependencies on it, so it might disappear. 

Other than that, everything seems fine.

**Comment by db0** [comment]
OK I'm merging now but if you find a more elegant approach later, feel free to improve

**Comment by ninele7** [comment]
Do you plan to do any announcements, to get workers to update to newer code and runtime, so whole horde would run faster?

**Comment by db0** [comment]
I did ping all the workers. No need for an announcement for everyone for that
